feat: add AccessControlEnumerable extension

klingonaston opened the pull request:
<!--
Thank you for your interest in contributing to OpenZeppelin!

Consider opening an issue for discussion prior to submitting a PR. New features will be merged faster if they were first discussed and designed with the team.

Describe the changes introduced in this pull request. Include any context necessary for understanding the PR's purpose.
-->
### Add AccessControlEnumerable Extension

- Introduced `enumerable` module for enhanced access control functionality that allows enumerating members of each role
- Implemented OpenZeppelin's AccessControlEnumerable pattern in Rust/Stylus
- Updated public exports to include new access control components

<!-- Fill in with issue number -->
Resolves #361 

#### PR Checklist

<!--
Before merging the pull request all of the following must be completed.
Feel free to submit a PR or Draft PR even if some items are pending.
Some of the items may not apply.
-->

- [x] Tests
- [x] Documentation
- [x] Changelog

klingonaston commented:
Hey @bidzyyys , I've opened a PR.  when you get a chance could you review it and share any feedback?

bidzyyys reviewed (commented):
Hey @klingonaston!
Thanks for submitting PR.
Please at first:
- Improve Rust docs to match our current attitude
- Add an example for `AccessControlEnumerable` as we did e.g. for `Erc20Wrapper` (E2E tests only happy paths).

klingonaston commented:
Hi @bidzyyys , thanks a lot for your suggestions!
I've updated the Rust docs and added an example for AccessControlEnumerable. could you check if I'm doing what you asked for? If anything seems off or not quite right, I'd really appreciate your guidance on how to correct it. Thanks again!

klingonaston commented:
Hi @bidzyyys , just following up on this. Would appreciate it if you could take a look at the PR when you have some time. Happy to address any feedback. Thanks!

bidzyyys commented on contracts/src/access/enumerable.rs:
```diff
@@ -0,0 +1,330 @@
+pub struct AccessControlEnumerable {
+    /// Base access control functionality
+    #[borrow]
+    access: AccessControl,
```
Hey @klingonaston!
Instead of having `AccessControl` as an attribute, please try to achieve the same design as we have for [Erc721 Enumerable extension](https://github.com/OpenZeppelin/rust-contracts-stylus/blob/main/contracts/src/token/erc721/extensions/enumerable.rs) -- reference to `AccessControl` contract  should be passed to `AccessControlEnumerable` methods as you can see in [Erc721 Enumerable example](https://github.com/OpenZeppelin/rust-contracts-stylus/blob/main/examples/erc721/src/lib.rs).

bidzyyys reviewed (commented):
Please redesign the extension to be similar to [Erc721 Enumerable](https://github.com/OpenZeppelin/rust-contracts-stylus/blob/main/contracts/src/token/erc721/extensions/enumerable.rs) and fix all the failing CI jobs.

bidzyyys commented:
Hey @klingonaston!
Sorry for the delay, but we were struggling with publishing `v0.2.0-rc.0` release. 

Are you still interested in finishing and merging your work into `main` branch?

bidzyyys commented on contracts/src/access/mod.rs:
Cleaning needed here.

bidzyyys commented on contracts/src/access/enumerable.rs:
```diff
@@ -0,0 +1,330 @@
+    enumeration: RoleEnumeration,
+}
+
+#[external]
```
Should be consistent with the new inheritance model.

bidzyyys commented on examples/access-control-enumerable/Cargo.toml:
Make sure it is consistent with other examples.

bidzyyys commented on examples/access-control-enumerable/src/constructor.sol:
Not needed thanks to #639 !

bidzyyys reviewed (commented):
Missing Antora docs and refactor to the new inheritance model.

bidzyyys commented on examples/access-control/tests/abi/mod.rs:
```diff
@@ -1,6 +1,7 @@
 #![allow(dead_code)]
 use alloy::sol;
 
+// TODO: add AccessControlEnumerable interface.
```
```suggestion
```

bidzyyys commented on examples/access-control/tests/access_control.rs:
```diff
@@ -518,3 +587,5 @@ async fn error_when_previous_admin_revokes_roles(
     Ok(())
 }
+
+// TODO: add AccessControlEnumerable E2E tests.
```
```suggestion
```

bidzyyys reviewed (approved):
LGTM

qalisander commented on contracts/src/utils/structs/enumerable_set.rs:
```diff
@@ -0,0 +1,313 @@
+        // storage slot.
+        let position = self.positions.get(value);
+
+        if position.is_zero() {
```
It can be refactored a bit to return early, but I don't mind to keep it similar to solidity

qalisander reviewed (approved):
LGTM!

bidzyyys commented:
@0xNeshi I will create a GH Issue about making `EnumerableSet` generic (not only for `Address` type).

0xNeshi commented on contracts/src/utils/structs/enumerable_set.rs:
```diff
@@ -0,0 +1,313 @@
+    ///
+    /// * `&mut self` - Write access to the set's state.
+    /// * `value` - The value to remove from the set.
+    ///
+    /// # Panics
+    ///
+    /// * Should never panic.
```
```suggestion
```
nit: we agreed to have panics section only when we expect panics to happen.

0xNeshi commented on contracts/src/utils/structs/enumerable_set.rs:
```diff
@@ -0,0 +1,313 @@
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the set's state.
+    ///
+    /// # Panics
+    ///
+    /// * Should never panic.
```
```suggestion
```

0xNeshi commented on contracts/src/utils/structs/enumerable_set.rs:
```diff
@@ -0,0 +1,313 @@
+    ///
+    /// * Should never panic.
+    pub fn values(&self) -> Vec<Address> {
+        let mut values = Vec::new();
+        for idx in 0..self.values.len() {
+            values.push(
+                self.values
+                    .get(idx)
+                    .expect("element at index: {idx} must exist"),
+            );
+        }
+        values
```
Would be much more elegant if it were possible to have an iterator on `StorageVec`. 
Created an [issue on the SDK](https://github.com/OffchainLabs/stylus-sdk-rs/issues/273), hopefully it can be implemented.

0xNeshi commented on contracts/src/access/control/extensions/enumerable.rs:
```diff
@@ -0,0 +1,512 @@
+            Ok(())
+        }
+
+        fn revoke_role(
+            &mut self,
+            role: B256,
+            account: Address,
+        ) -> Result<(), Self::Error> {
+            let admin_role = self.get_role_admin(role);
+            self.only_role(admin_role)?;
+            self.enumerable._revoke_role(
+                role,
+                account,
+                &mut self.access_control,
+            );
+            Ok(())
+        }
```
One of the annoying limitations of Stylus's inheritance model is that it makes it difficult, if not impossible, to write legit unit tests.
Devs are forced to write something like this, which looks more like an e2e test than a unit test...

0xNeshi reviewed (approved):
Looks good, left nits
